src/twisted/conch/ssh/session.py

```diff
@@ -180,16 +180,16 @@ def extReceived(self, dataType, data):
             log.warn("Weird extended data: {dataType}", dataType=dataType)
 
     def eofReceived(self):
+        if self.client:
+            self.conn.sendClose(self)
         if self.session:
             self.session.eofReceived()
-        elif self.client:
-            self.conn.sendClose(self)
 
     def closed(self):
+        if self.client and self.client.transport:
+            self.client.transport.loseConnection()
```

Comment:
Yeah, the problem is that the session adapter can't really gain a link to the client protocol without significant backward-compatibility difficulties, since the session adapter is entirely provided by the application; so I think we do need these as separate concepts, although as mentioned above maybe the client protocol could gain a link to the session adapter and then it would be able to deal with notifying the session adapter.

Notifying the actual SSH client of the closure is dealt with by the `self.conn.sendClose(self)` call in `SSHSession.eofReceived`, which sends `MSG_CHANNEL_CLOSE` on the wire.